### PR TITLE
Fix mypy error in pandas/tests.indexes.test_base.py

### PR DESCRIPTION
**Pull request by rushabh-v**
- [x] relates to #28926 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by WillAyd** [comment]
What error does this give?

**Comment by RushabhVasani** [comment]
Hello @WillAyd 
Thanks for responding.
It was giving error: Need type annotation for 'x'.

**Review by WillAyd — APPROVED** [review]
OK lgtm. Can you merge master one more time and push? Fixed some things with CI recently

@simonjayhawkins

**Comment by RushabhVasani** [comment]
Hello @WillAyd 
I am actually a bit new to git commands. And I have made this pull request using the UI way.
I have no idea how to merge master again in the same pull request.
Can you please help me out.
Do I have to clone my forked repo and then run 'git rebase' and 'git push' or it is something else.
Thank you.

**Comment by RushabhVasani** [comment]
@WillAyd 
Can I just add an extra space somewhere commit and in second commit remove that extra space in order to achieve this.
I am sorry if it's a wrong way.
Thanks!

**Comment by RushabhVasani** [comment]
I have merged it again. Please review the changes.
 @WillAyd  @simonjayhawkins

**Comment by RushabhVasani** [comment]
I have merged it again. Please review the changes.
@WillAyd  @simonjayhawkins

**Comment by simonjayhawkins** [comment]
tests are failing see https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=19636

**Comment by RushabhVasani** [comment]
I have started working on them.

**Comment by RushabhVasani** [comment]
Hello @simonjayhawkins @WillAyd 
It is giving the error: Slice index must be an integer or None.
This error is coming from mypy.  And I think there is a version conflict. Because this bug is solved with https://github.com/python/mypy/issues/2410.
The lines giving this error are https://github.com/RushabhVasani/pandas/blob/fcddcfcda2de3306191067afe19bc86859665091/pandas/tests/indexes/test_base.py#L1741-L1756
So I tried to run the same code on my local machine. And it all works fine. The code I run was,
 `>>> pytest.mark.parametrize(
...         "in_slice,expected",
...         [
...             (pd.IndexSlice[::-1], "yxdcb"),
...             (pd.IndexSlice["b":"y":-1], ""),
...             (pd.IndexSlice["b"::-1], "b"),
...             (pd.IndexSlice[:"b":-1], "yxdcb"),
...             (pd.IndexSlice[:"y":-1], "y"),
...             (pd.IndexSlice["y"::-1], "yxdcb"),
...             (pd.IndexSlice["y"::-4], "yb"),
...             # absent labels
...             (pd.IndexSlice[:"a":-1], "yxdcb"),
...             (pd.IndexSlice[:"a":-2], "ydb"),
...             (pd.IndexSlice["z"::-1], "yxdcb"),
...             (pd.IndexSlice["z"::-3], "yc"),
...             (pd.IndexSlice["m"::-1], "dcb"),
...             (pd.IndexSlice[:"m":-1], "yx"),
...             (pd.IndexSlice["a":"a":-1], ""),
...             (pd.IndexSlice["z":"z":-1], ""),
...             (pd.IndexSlice["m":"m":-1], ""),
...         ],
...     )`
And the output it gave was,

MarkDecorator(mark=Mark(name='parametrize', args=('in_slice,expected', [(slice(None, None, -1), 'yxdcb'), (slice('b', 'y', -1), ''), (slice('b', None, -1), 'b'), (slice(None, 'b', -1), 'yxdcb'), (slice(None, 'y', -1), 'y'), (slice('y', None, -1), 'yxdcb'), (slice('y', None, -4), 'yb'), (slice(None, 'a', -1), 'yxdcb'), (slice(None, 'a', -2), 'ydb'), (slice('z', None, -1), 'yxdcb'), (slice('z', None, -3), 'yc'), (slice('m', None, -1), 'dcb'), (slice(None, 'm', -1), 'yx'), (slice('a', 'a', -1), ''), (slice('z', 'z', -1), ''), (slice('m', 'm', -1), '')]), kwargs={}))
>>>\

**Comment by RushabhVasani** [comment]
Hello @simonjayhawkins @WillAyd 
It is giving the error: Slice index must be an integer or None.
This error is coming from mypy packages.  And I think there is a mypy version conflict. Because this bug is solved with (https://github.com/python/mypy/issues/2410)
The lines giving this error are (https://github.com/RushabhVasani/pandas/blob/fcddcfcda2de3306191067afe19bc86859665091/pandas/tests/indexes/test_base.py#L1741-L1756)
So I tried to run the same code on my local machine. And it all works fine. The code I run was,

 

```
@pytest.mark.parametrize( 
       "in_slice,expected",
        [
            (pd.IndexSlice[::-1], "yxdcb"),
            (pd.IndexSlice["b":"y":-1], ""),
            (pd.IndexSlice["b"::-1], "b"),
            (pd.IndexSlice[:"b":-1], "yxdcb"),
            (pd.IndexSlice[:"y":-1], "y"),
            (pd.IndexSlice["y"::-1], "yxdcb"),
            (pd.IndexSlice["y"::-4], "yb"),
            # absent labels
            (pd.IndexSlice[:"a":-1], "yxdcb"),
            (pd.IndexSlice[:"a":-2], "ydb"),
            (pd.IndexSlice["z"::-1], "yxdcb"),
            (pd.IndexSlice["z"::-3], "yc"),
            (pd.IndexSlice["m"::-1], "dcb"),
            (pd.IndexSlice[:"m":-1], "yx"),
            (pd.IndexSlice["a":"a":-1], ""),
            (pd.IndexSlice["z":"z":-1], ""),
            (pd.IndexSlice["m":"m":-1], ""),
        ],
    )
```
And the output it gave was,

`MarkDecorator(mark=Mark(name='parametrize', args=('in_slice,expected', [(slice(None, None, -1), 'yxdcb'), (slice('b', 'y', -1), ''), (slice('b', None, -1), 'b'), (slice(None, 'b', -1), 'yxdcb'), (slice(None, 'y', -1), 'y'), (slice('y', None, -1), 'yxdcb'), (slice('y', None, -4), 'yb'), (slice(None, 'a', -1), 'yxdcb'), (slice(None, 'a', -2), 'ydb'), (slice('z', None, -1), 'yxdcb'), (slice('z', None, -3), 'yc'), (slice('m', None, -1), 'dcb'), (slice(None, 'm', -1), 'yx'), (slice('a', 'a', -1), ''), (slice('z', 'z', -1), ''), (slice('m', 'm', -1), '')]), kwargs={}))`

So I think there is a mypy version conflict.

**Comment by simonjayhawkins** [comment]
> So I think there is a mypy version conflict.

yes these errors are not reported in 0.730. It was a typeshed fix though, see https://github.com/python/typeshed/pull/3024 which was pulled into mypy for version 0.730

we are currently pinned to 0.720 on ci.

**Comment by rushabh-v** [comment]
>we are currently pinned to 0.720 on ci.

Okay, then what should be done now.

**Comment by simonjayhawkins** [comment]
> > we are currently pinned to 0.720 on ci.
> 
> Okay, then what should be done now.

I guess two options.

1. seperate PR to bump mypy to 0.730 as a precursor to this
2. add `# type: ignore` with comments for error message and link to typeshed issue. 

eg.

https://github.com/pandas-dev/pandas/blob/2ca216101d03af76be3f5a4d6447571f0628fbb1/pandas/io/formats/printing.py#L490-L494

@WillAyd any preference?

**Comment by simonjayhawkins** [comment]
@rushabh-v can you merge master. mypy has been bumped to 0.730 on ci.

**Comment by rushabh-v** [comment]
It is giving the same error(Slice index must be an integer or None) again.
@simonjayhawkins

**Comment by WillAyd** [comment]
Yea I see the same locally - can you investigate further @rushabh-v ?

**Comment by rushabh-v** [comment]
Yes, I am looking into it. @WillAyd

**Comment by rushabh-v** [comment]
I am sorry, I mistakenly committed directly to master.

**Comment by rushabh-v** [comment]
I am trying to use mypy version 0.740 in a different branch but when I change mypy version in environment.yml and requirements-dev.txt It starts giving [this](https://github.com/rushabh-v/pandas/commit/d147286a6f3f9a9109fa7d041e7e3504752b3118/checks?check_suite_id=318002155l) error.

I noticed that requirements-dev.txt should be generated from environment.yml so created a new branch and then tried changing only environment.yml but that also gave the same error.

**Comment by rushabh-v** [comment]
It is giving this error for every version of mypy on my local machine.  
> I guess two options.
>1. seperate PR to bump mypy to 0.730 as a precursor to this
>>2. add # type: ignore with comments for error message and link to typeshed issue.
 Should we go with the 2nd option? @simonjayhawkins  @WillAyd 
I have checked locally adding #type: ignore works fine on mypy 0.730.

**Comment by rushabh-v** [comment]
It is giving this error for every version of mypy on my local machine.

>I guess two options.
>seperate PR to bump mypy to 0.730 as a precursor to this
>>add # type: ignore with comments for error message and link to typeshed issue.

Should we go with the 2nd option? @simonjayhawkins @WillAyd
I have checked locally, adding #type: ignore works fine on mypy 0.730.

**Comment by rushabh-v** [comment]
It is giving this error for every version of mypy on my local machine.
>I guess two options.
>1. seperate PR to bump mypy to 0.730 as a precursor to this
>>2. add # type: ignore with comments for error message and link to typeshed issue.

Should we go with the 2nd option? @simonjayhawkins @WillAyd
I have checked locally adding #type: ignore works fine on mypy 0.730.

**Comment by simonjayhawkins** [comment]
replacing `(pd.IndexSlice["b":"y":-1], ""),` with `(pd.IndexSlice[slice("b", "y", -1)], ""),` silences mypy. so it appears that the current mypy errors are not directly related to/ been solved by https://github.com/python/typeshed/pull/3024

the mypy issue, https://github.com/python/mypy/issues/2410 is still open, so a `# type: ignore` with a comment referencing the mypy issue is probably the most practical solution here.

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -1732,23 +1732,23 @@ def test_slice_locs_na_raises(self):
...
         "in_slice,expected",
         [
-            (pd.IndexSlice[::-1], "yxdcb"), # type: ignore 
-            (pd.IndexSlice["b":"y":-1], ""), # type: ignore 
-            (pd.IndexSlice["b"::-1], "b"), # type: ignore 
-            (pd.IndexSlice[:"b":-1], "yxdcb"), # type: ignore 
-            (pd.IndexSlice[:"y":-1], "y"), # type: ignore
-            (pd.IndexSlice["y"::-1], "yxdcb"), # type: ignore 
-            (pd.IndexSlice["y"::-4], "yb"), # type: ignore 
+            (pd.IndexSlice[::-1], "yxdcb"),  # type: ignore
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/test_base.py]: the ignore shouldn't be needed on this line.

run mypy with `--warn-unused-ignores` to confirm.

Comment (rushabh-v) [reply pandas/tests/indexes/test_base.py]: >the ignore shouldn't be needed on this line.
yes, that's true! I am removing it.

**Review by simonjayhawkins — APPROVED** [review]
@rushabh-v Thanks for sticking with this. @WillAyd

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @rushabh-v

**Pull request merged**